### PR TITLE
feat: change `any` to something more meaningful

### LINKED ISSUE #164: SDK - Change `any` types to something meaningful

### PR DESCRIPTION
**Pull request by flechtnergalani**
@Dudleyneedham 
Squashed commits from #196 on top of develop and resolved merge conflicts, then added two more commits replacing some more `any`s. 

fixes #164 

- defined any typed where they could be narrowed down to more specific types
- refactorings to avoid using any return or argument types
- removed await on non-async function calls

## How to test:

- [x] run unit tests
- [x] run integration tests
- [x] test with demo client 

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: src/did/Did.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export interface IDidDocumentPublicKey {
 export interface IDidDocumentPpties {
   authentication: object
   publicKey: IDidDocumentPublicKey[]
-  service: any
+  service: any // can this possibly be narrowed down to something other than any?
```
Comment (flechtnergalani) [inline src/did/Did.ts]: @maudnals I'll need your input here - `any` ideas? Please?

File: src/ctype/CType.spec.ts [hunk 2]
```diff
@@ -63,7 +63,7 @@ describe('CType', () => {
       owner: identityAlice.address,
     }
 
-    const resultTxStatus = new TxStatus(FINALIZED, Crypto.hashStr('987654'))
```
Comment (Dudleyneedham) [inline src/ctype/CType.spec.ts]: This is out of the scope of the ticket.

File: src/delegation/Delegation.chain.ts [hunk 3]
```diff
@@ -37,7 +37,10 @@ export async function getChildIds(
   id: IDelegationBaseNode['id']
 ): Promise<string[]> {
   const blockchain = await getCached()
-  return Blockchain.asArray(await blockchain.api.query.delegation.children(id))
```
Comment (Dudleyneedham) [inline src/delegation/Delegation.chain.ts]: This is out of the scope of the ticket.

File: src/delegation/DelegationNode.chain.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ export async function store(
   const includeParentId: boolean = delegation.parentId
     ? delegation.parentId !== delegation.rootId
     : false
-  const tx: SubmittableExtrinsic = await blockchain.api.tx.delegation.addDelegation(
```
Comment (Dudleyneedham) [inline src/delegation/DelegationNode.chain.ts]: See other comments about the ticket.

File: src/messaging/Message.ts [hunk 5]
```diff
@@ -281,7 +281,7 @@ export interface IRequestAcceptDelegation extends IMessageBodyBase {
       isPCR: boolean
     }
     metaData?: {
-      [key: string]: any
+      [key: string]: string | object // can this possibly be narrowed down to something other than any?
```
Comment (Dudleyneedham) [inline src/messaging/Message.ts]: You will need to remove the comments.

File: src/quote/Quote.ts [hunk 6]
```diff
@@ -29,8 +29,10 @@ export function validateQuoteSchema(
   )
   if (!result && ajv.errors) {
     if (messages) {
-      ajv.errors.forEach((error: any) => {
-        messages.push(error.message)
+      ajv.errors.forEach((error: Ajv.ErrorObject) => {
```
Comment (Dudleyneedham) [inline src/quote/Quote.ts]: Nice

File: src/types/CType.ts [hunk 7]
```diff
@@ -5,9 +5,9 @@
 import IPublicIdentity from './PublicIdentity'
 
 export interface ICTypeSchema {
-  $id: any
-  $schema: any
-  properties: any
+  $id: string
+  $schema: string
+  properties: object // TO DO: need to refine what properties are
```
Comment (Dudleyneedham) [inline src/types/CType.ts]: This needs to be changed.

File: src/ctype/CTypeUtils.ts [hunk 8]
```diff
@@ -8,34 +8,41 @@ import Ajv from 'ajv'
...
-export function verifyClaimStructure(claim: any, schema: any): boolean {
+export function verifyClaimStructure(
+  claimContents: IClaim['contents'], // Checking the SDK and demo-client I feel this should be named claimContents
+  schema: ICType['schema']
+): boolean {
   if (!verifySchema(schema, CTypeModel)) {
     throw new Error('CType does not correspond to schema')
   }
-  return verifySchema(claim, schema)
+  return verifySchema(claimContents, schema)
```
Comment (Dudleyneedham) [inline src/ctype/CTypeUtils.ts]: I agree this needs to be changed to this. But, it is out of the scope of the ticket.

File: src/ctype/CType.spec.ts (hunk 2 above)
Comment (flechtnergalani) [reply src/ctype/CType.spec.ts]: A change will be necessary though if we’re changing the type of TxStatus.payload to object. We could pass an object instead, but this is dead code anyway, it is never used afterwards.

File: src/delegation/Delegation.chain.ts (hunk 3 above)
Comment (flechtnergalani) [reply src/delegation/Delegation.chain.ts]: I wouldn't think it is - it's a necessary change to get rid of the return type `any[]` of `Blockchain.asArray()` 
What we probably could do instead is convert each element to string in order to match the `string[]` return type of `getChildIds()`

File: src/delegation/DelegationNode.chain.ts (hunk 4 above)
Comment (flechtnergalani) [reply src/delegation/DelegationNode.chain.ts]: I agree this could be merged separately

**Comment by weichweich** [comment]
https://github.com/KILTprotocol/sdk-js/blob/2dde442d0e87d948f08638ed83120d5390364a28/src/types/RequestForAttestation.ts#L20

We could add a type for that. 🙂 
```ts
export type NonceHashTree = NonceHash | { [key: string]: NonceHashTree }
```

**Comment by flechtnergalani** [comment]
> https://github.com/KILTprotocol/sdk-js/blob/2dde442d0e87d948f08638ed83120d5390364a28/src/types/RequestForAttestation.ts#L20
> 
> We could add a type for that. 🙂
> 
> ```ts
> export type NonceHashTree = NonceHash | { [key: string]: NonceHashTree }
> ```

I like that ! But https://github.com/KILTprotocol/sdk-js/blob/2dde442d0e87d948f08638ed83120d5390364a28/src/requestforattestation/RequestForAttestation.ts#L308 would technically require it to be 
```ts
export type NonceHashTree = { [key: string]: NonceHash }
```
Would also work!

**Comment by weichweich** [comment]
> ```ts
> export type NonceHashTree = { [key: string]: NonceHash }
> ```
> Would also work!

But that's not a tree. If we want to be able to hide arbitrarily nested fields we need a tree (see KILTprotocol/ticket/issues/386 for details).

**Comment by weichweich** [comment]
https://github.com/KILTprotocol/sdk-js/blob/2dde442d0e87d948f08638ed83120d5390364a28/src/types/Claim.ts#L10

could also have this type, which is pretty close to `any`. 🤣 

```ts
export type ClaimContent =
  | string
  | number
  | null
  | boolean
  | { [key: string]: ClaimContent }
  | ClaimContent[]
```

**Comment by flechtnergalani** [comment]
> ```ts
> ```ts
> export type NonceHashTree = { [key: string]: NonceHash }
> ```
> 
> 
> Would also work!
> 
> But that's not a tree. If we want to be able to hide arbitrarily nested fields we need a tree (see KILTprotocol/ticket/issues/386 for details).
> ```

It is not, but I'd say we stick to this until the code supports an actual tree - that should be happening as part of the issue you opened, not as part of this PR

**Comment by flechtnergalani** [comment]
> https://github.com/KILTprotocol/sdk-js/blob/2dde442d0e87d948f08638ed83120d5390364a28/src/types/Claim.ts#L10
> 
> could also have this type, which is pretty close to `any`. 🤣
> 
> ```ts
> export type ClaimContent =
>   | string
>   | number
>   | null
>   | boolean
>   | { [key: string]: ClaimContent }
>   | ClaimContent[]
> ```

An issue similar to the ClaimHashTree: It can't be a scalar or Array because the property definitions in the Ctype require a Struct / Object, so maybe
```ts
[...]
contents: { [key: string]: ClaimProperty }
[...]
export type ClaimProperty =
  | string
  | number
  | null
  | boolean
  | { [key: string]: ClaimProperty }
  | ClaimProperty[]
```
But it seems like the `CTypeModel` currently only allows four types: 
```ts
          properties: {
            type: {
              type: 'string',
              enum: ['string', 'integer', 'number', 'boolean'],
            },
```
in which case 
```ts
contents: { [key: string]: string | number | boolean }
```
covers it all - it doesn't even allow `null` when validating a claim against a ctype.

File: src/did/Did.ts [hunk 9]
```diff
@@ -67,7 +68,14 @@ export interface IDidDocumentPublicKey {
...
-  service: any
+  service: IDidService[]
+}
+
+export interface IDidService {
+  type: string
+  serviceEndpoint: string | AnyJsonObject
+  id?: string
+  [key: string]: AnyJson
 }
```
Comment (flechtnergalani) [inline src/did/Did.ts]: we implemented this on the basis of the [W3C draft](https://www.w3.org/TR/did-core/#service-endpoints), hope that makes sense to you @maudnals

File: src/ctype/CTypeUtils.ts (hunk 8 above)
Comment (flechtnergalani) [reply src/ctype/CTypeUtils.ts]: given that the name of the parameter should not have any consequences for the function call, can we not just leave it like this?

**Pull request merged**